client/a8c-for-agencies/sections/marketplace/products-overview/product-listing/hooks/use-product-and-plans.tsx

```diff
@@ -0,0 +1,188 @@
+	const products = getProductsAndPlansByFilter(
+		PRODUCT_FILTER_PRODUCTS,
+		filteredProductsAndBundles
+	);
+	const filteredProducts = MERGABLE_PRODUCTS.map( ( filter ) => {
+		return products.filter( ( { slug } ) => slug.startsWith( filter ) );
+	} )
+		.filter( ( subArray ) => subArray.length > 0 ) // Remove empty arrays
+		.map( ( mergedProducts ) =>
+			mergedProducts.length === 1 ? mergedProducts[ 0 ] : mergedProducts
+		); // flat out if only one product.
+
+	const restOfProducts = products.filter( ( { slug } ) => {
+		return ! MERGABLE_PRODUCTS.some( ( filter ) => slug.startsWith( filter ) );
+	} );
+
+	return [ ...restOfProducts, ...filteredProducts ].sort( ( a, b ) => {
+		const product_a = Array.isArray( a ) ? a[ 0 ].name : a.name;
+		const product_b = Array.isArray( b ) ? b[ 0 ].name : b.name;
+		return product_a.localeCompare( product_b );
+	} ) as APIProductFamilyProduct[];
+};
+
+export default function useProductAndPlans( {
+	selectedBundleSize = 1,
+	selectedSite,
+	selectedProductFilter = PRODUCT_FILTER_ALL,
+	productSearchQuery,
+	usePublicQuery = false,
+}: Props ) {
+	const { data, isLoading: isLoadingProducts } = useProductsQuery( usePublicQuery );
+
+	const addedPlanAndProducts = useSelector( ( state ) =>
+		selectedSite ? getAssignedPlanAndProductIDsForSite( state, selectedSite.ID ) : null
+	);
+
+	return useMemo( () => {
+		// List only products that is compatible with current bundle size.
+		const supportedProducts =
+			selectedBundleSize > 1
+				? data?.filter(
+						( { supported_bundles } ) =>
+							supported_bundles?.some?.( ( { quantity } ) => selectedBundleSize === quantity )
+				  )
+				: data;
+
+		// We pre-filter the list by current selected filter
+		let filteredProductsAndBundles = getProductsAndPlansByFilter(
+			selectedProductFilter,
+			supportedProducts
+		);
+
+		// Filter products based on the search term
+		if ( productSearchQuery ) {
+			filteredProductsAndBundles = filteredProductsAndBundles.filter( ( product ) =>
+				isProductMatch( product, productSearchQuery )
+			);
+		}
+
+		// Filter products & plan that are already assigned to a site
+		if ( selectedSite && addedPlanAndProducts && filteredProductsAndBundles ) {
+			filteredProductsAndBundles = filteredProductsAndBundles.filter(
+				( product ) => ! addedPlanAndProducts.includes( product.product_id )
+			);
+		}
+
+		// We need the suggested products (i.e., the products chosen from the dashboard) to properly
+		// track if the user purchases a different set of products.
+		const suggestedProductSlugs = getQueryArg( window.location.href, 'product_slug' )
+			?.toString()
+			.split( ',' );
+
+		return {
+			isLoadingProducts,
+			data,
+			filteredProductsAndBundles,
+			plans: getDisplayablePlans( filteredProductsAndBundles ),
+			products: getDisplayableProducts( filteredProductsAndBundles ),
+			backupAddons: getProductsAndPlansByFilter(
+				PRODUCT_FILTER_VAULTPRESS_BACKUP_ADDONS,
+				filteredProductsAndBundles
+			),
```

Comment:
[non-blocker] This is not needed for now, but we will probably need to update the field names here to be more Jetpack-specific.

Reply:
Sounds good 👍